stdlib/public/core/StringComparison.swift

```diff
@@ -0,0 +1,1368 @@
+//
+internal func _findNormalizationSegmentEnd(
+  _ buf: UnsafeBufferPointer<UInt16>,
+  startingFrom idx: Int
+) -> Int {
+  let count = buf.count
+  _sanityCheck(idx < count, "out of bounds")
+
+  // Normalization boundaries are best queried before known starters. Advance
+  // past one scalar first.
+  var (_, segmentEndIdx) = _parseRawScalar(buf, startingFrom: idx)
+  while segmentEndIdx < count {
+    let (scalar, nextIdx) = _parseRawScalar(buf, startingFrom: segmentEndIdx)
+    if scalar._hasNormalizationBoundaryBefore {
+      break
+    }
+    segmentEndIdx = nextIdx
+  }
+  return segmentEndIdx
+}
+internal func _findNormalizationSegmentStart(
+  _ buf: UnsafeBufferPointer<UInt16>,
+  endingAt idx: Int // one-past-the-end
+) -> Int {
+  var idx = idx
+  let count = buf.count
+  _sanityCheck(idx > 0 && idx <= count, "out of bounds")
+
+  while idx > 0 {
+    let (scalar, priorIdx) = _reverseParseRawScalar(buf, endingAt: idx)
+    idx = priorIdx
+    if scalar._hasNormalizationBoundaryBefore {
+      break
+    }
+  }
+  return idx
+}
+
+internal func _findNormalizationSegment(
+  _ buf: UnsafeBufferPointer<UInt16>,
+  spanning idx: Int
+) -> (Int, Int) {
+  let ptr = buf.baseAddress._unsafelyUnwrappedUnchecked
+  var idx = idx
+
+  // Corner case: if we're sub-surrogate, back up
+  if _slowPath(
+    idx > 0 && _isTrailingSurrogate(ptr[idx]) && _isLeadingSurrogate(ptr[idx-1])
+  ) {
+    idx -= 1
+  }
+  let segmentEnd = _findNormalizationSegmentEnd(buf, startingFrom: idx)
+
+  // Find the start
+  if _slowPath(idx == 0) {
+    return (0, segmentEnd)
+  }
+
+  // Check current scalar
+  if _parseRawScalar(buf, startingFrom: idx).0._hasNormalizationBoundaryBefore {
+    return (idx, segmentEnd)
+  }
+
+  // Reverse parse until we found the segment start
+  let segmentStart = _findNormalizationSegmentStart(buf, endingAt: idx)
+
+  return (segmentStart, segmentEnd)
+}
+
+// Wether the segment identified by `idx` is prenormal.
+//
+// Scalar values below 0x300 are special: normalization segments containing only
+// one such scalar are trivially prenormal under NFC. Most Latin-derived scripts
+// can be represented entirely by <0x300 scalar values, meaning that many user
+// strings satisfy this prenormal check. We call sub-0x300 scalars "Latiny" (not
+// official terminology).
+//
+// The check is effectively:
+//   1) Whether the current scalar <0x300, AND
+//   2) Whether the current scalar comprises the entire segment
+//
+internal func _isLatinyPrenormal(
+  _ buffer: UnsafeBufferPointer<UInt16>, idx: Int
+) -> Bool {
+  _sanityCheck(idx < buffer.count, "out of bounds")
+  let ptr = buffer.baseAddress._unsafelyUnwrappedUnchecked
+
+  let cu = ptr[idx]
+  if _slowPath(cu >= 0x300) {
+    return false
+  }
+  if _slowPath(idx+1 == buffer.count) {
+    return true
+  }
+
+  let nextCU = ptr[idx+1]
+  return nextCU < 0x300 || _hasNormalizationBoundary(before: nextCU)
+}
+
+public func _compareStringsPreLoop(
```

Comment:
All of these global functions can now be on the relevant `_UnmanagedString where CodeUnit == UIntX` structures.